### ISSUE TITLE
C API: Consider using the same type for length and indices

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1271]
Currently, `duk_get_length()` returns a value of type `size_t`, while e.g. `duk_get_prop_index()` takes a value of type `duk_uarridx_t`, which usually maps to `unsigned int`.  For code that must iterate through an array, this leads to extraneous casting between the two types to avoid compiler warnings.

**Comment by svaarala** [comment]
Not only that, but the types may have different sizes. For example, `duk_uarridx_t` may be 32 bits but `duk_size_t` (which currently maps to `size_t` as is) may be only 16 bits.

The difficulty with `duk_get_length()` is that it returns a relevant length for multiple types whose logical types are different:
- For strings and buffers `size_t` is currently useful because it corresponds with the maximum allocation size, and since there's no virtualization of indices, and actual realized length is unlikely to be larger than the `size_t` type. Also, `size_t` matches C code expectations for a memory pointer/length pair. In ES2015+ string/buffer size can be `2^53^ so a `duk_uint_t` is not enough.
- For arrays a guaranteed unsigned 32-bit integer would be needed. Arrays can be sparse so they may be larger than maximum allocation size. Technically `size_t` can be 16 bits and even in that case it's possible for an array to have length of 2^32-1.
- Any other object: looks up `.length` and coerces to integer. For this there's no real limit (except IEEE) what `.length`could be.

While it's convenient you can use the same API call for multiple types, it doesn't actually currently work well from a typing perspective. To improve, the return type could be changed, but I'm not sure there's a guaranteed "supertype" other than an IEEE double that can always cover the different return values.

Another option is to split up the API for different types so that the C type matches the script type better. For example, it's quite logical to use `size_t` in combination with memory pointers because that's the common thing to do in C code, but doing so for non-memory-related lengths seems odd.

One more option is to rework the C types somehow, but I haven't come up with a good idea so far. For example:
- Change `duk_size_t` so that it is guaranteed to contain an array index. Easy to do, but means `duk_size_t` may be longer than `size_t`, and any calls into e.g. `memcpy()` will need a check/cast unlike now.
- Change `duk_uint_t` so that it can contain both an array index and string/buffer length. With ES2015+ requires a 64-bit type which can be very inefficient. So in practice the APIs dealing with virtual integer indices and their related lengths may need a specific type beyond `duk_uint_t`.
- In general start using a smaller set of "compromise types". For example, use `duk_uint_t` in place of `size_t`. That would force `duk_uint_t` to be 64 bits on 64-bit platforms however which is currently not necessary (and sometimes leads to better code). Using `duk_uint_t` in place of `duk_uarridx_t` would be fine for now, but if array indices get >4G support then that becomes problematic.

**Comment by fatcerberus** [comment]
It was my understanding that ES6 also specified 2^53-1 range for array indices as well?  So the requirements for `duk_uarridx_t` actually match those of string/buffer `.length`, it seems.

**Comment by svaarala** [comment]
That's not yet the case:

> An integer index is a String-valued property key that is a canonical numeric String (see 7.1.16) and whose numeric value is either +0 or a positive integer ≤ 2^53−1. An array index is an integer index whose numeric value i is in the range +0 ≤ i < 2^32−1.

But it would be nice to anticipate it growing beyond 4G.

**Comment by fatcerberus** [comment]
Ah, thanks.  The question then is, should `duk_get/set_prop_index()` take an array index specifically for its argument, or an "integer index"? :)

I'll wait for that table though, I agree that will help guide the discussion better.

**Comment by fatcerberus** [comment]
Unfortunately, I think any fix made here will have to wait for Duktape 3.0.0 at the earliest because changing types is API incompatible.

**Comment by svaarala** [comment]
> Ah, thanks. The question then is, should duk_get/set_prop_index() take an array index specifically for its argument, or an "integer index"? :)

Right, and as discussed previously, a compliant "integer index" needs to be 64 bits which means it'll be inefficient or downright unavailable on some targets. So the C API may not be able to model the Ecmascript index space 1:1. For example, on a target without 64-bit types, it may be OK to restrict array, buffer, and string indices to 32 bits instead of the full 53 bits.

**Comment by fatcerberus** [comment]
> For example, on a target without 64-bit types, it may be OK to restrict array, buffer, and string indices to 32 bits instead of the full 53 bits.

No argument there.  I was just pointing it out to highlight that, from an API standpoint, it makes most sense for `duk_get_length()` and `duk_uarridx_t` to be the same underlying type.

**Comment by svaarala** [comment]
If `duk_get_length()` returned an "integer index" type, it might work out. That type would in practice never be smaller than `size_t`: it would either be 64-bit, or 32-bit on a platform where one wouldn't normally use 64-bit types.

The awkward case is a platform where 64-bit types are available but very inefficient. The typing would then lead to using 64-bit loop counters (unless one explicitly cast to a smaller type) if one simply declared the loop index / limit using the same type as returned by the API (which is quite natural).

**Comment by fatcerberus** [comment]
For what it's worth I usually end up just casting `duk_get_length()` to `duk_uarridx_t` anyway.  Otherwise I end up casting in the loop itself, which makes for ugly code.

**Comment by fatcerberus** [comment]
One option that would be convenient for more platforms is to use `size_t` as the base type for integer indices in the API.

**Comment by svaarala** [comment]
That or `ptrdiff_t` would be a pretty good choice in most cases, however:
- `size_t` may (technically) be only 16 bits. Though this would be very rare in practice, even on platforms with a 16-bit int.
- `size_t` may be 32 bits, which works OK for buffers and strings: even if the technical limit is 2^53 one can't allocate such a large buffer/string. But for virtual indexes (Proxy) or sparse arrays (in the future with >4G arrays) it's possible to have indices beyond 4G even while consuming a small amount of memory.
